src/middlewared/middlewared/plugins/vpn.py

```diff
@@ -0,0 +1,532 @@
+            ], {
+                'get': True
+            }
+        )
+        client_cert = await self.middleware.call(
+            'certificate.query', [
+                ['id', '=', client_certificate_id],
+                ['revoked', '=', False]
+            ]
+        )
+        if not client_cert:
+            raise CallError(
+                'Please provide a client certificate id for a certificate which exists on '
+                'the system and hasn\'t been marked as revoked.'
+            )
+        else:
+            client_cert = client_cert[0]
+            if (
+                await OpenVPN.common_validation(
+                    self.middleware, {
+                        **config,
+                        'client_certificate': client_certificate_id
+                    }, '', 'client'
+                )
+            )[0]:
+                raise CallError(
+                    'Please ensure provided client certificate exists in Root CA chain '
+                    'and has necessary extensions set.'
+                )
+
+        client_config = [
+            'client',
+            f'dev {config["device_type"].lower()}',
+            f'proto {config["protocol"].lower()}',
+            f'port {config["port"]}',
+            'remote "PLEASE FILL OUT SERVER DOMAIN/IP HERE"',
+            'user nobody',
+            'group nobody',
+            'persist-key',
+            'persist-tun',
+            '<ca>',
+            f'{root_ca["certificate"]}',
+            '</ca>',
+            '<cert>',
+            client_cert['certificate'],
+            '</cert>',
+            '<key>',
+            client_cert['privatekey'],
+            '</key>',
+            'verb 3',
+            'remote-cert-tls server',
+            f'compress {config["compression"].lower()}' if config['compression'] else None,
+            f'auth {config["authentication_algorithm"]}' if config['authentication_algorithm'] else None,
+            f'cipher {config["cipher"]}' if config['cipher'] else None,
+        ]
+
+        if config['tls_crypt_auth_enabled']:
+            client_config.extend([
+                '<tls-crypt>',
+                config['tls_crypt_auth'],
+                '</tls-crypt>'
+            ])
+
+        return (
+            '\n'.join(
+                filter(
+                    bool,
+                    client_config
+                )
+            ) + f'\n{config["additional_parameters"]}'
+        ).strip()
+
+    @accepts(
+        Dict(
+            'openvpn_server_update',
+            Bool('tls_crypt_auth_enabled'),
+            Int('netmask', validators=[Range(min=0, max=128)]),
+            Int('server_certificate'),
+            Int('port', validators=[Port()]),
+            Int('root_ca'),
+            IPAddr('server'),
+            Str('additional_parameters'),
+            Str('authentication_algorithm', enum=OpenVPN.digests(), null=True),
+            Str('cipher', null=True, enum=OpenVPN.ciphers()),
```

Comment:
These both seem to have tons of options.
It seems to make more sense to have `cipher_choices` and `authentication_algorithm_choices` instead.

Reply:
To be sure i understand correctly, are you referring to changing `OpenVPN.ciphers` and `OpenVPN.digests` names to the ones you mentioned ?

Reply:
No, remove the enums and have the choices methods in each service.